File: app/src/main/java/com/iamagamedev/mybrandnewgame/gameObjects/spells/SpellObject.java
package com.iamagamedev.mybrandnewgame.gameObjects.spells;

import android.util.Log;

import com.iamagamedev.mybrandnewgame.Constants.SpellNames;
import com.iamagamedev.mybrandnewgame.gameObjects.GameObject;

public abstract class SpellObject extends GameObject {

    private int waitingTime = 3000;
    private float x;
    private float y;

    public void updatePosition(float x, float y) {
        this.x = x;
        this.y = y;
        this.setWorldLocation(x, y, getWorldLocation().z);
    }

    @Override
    public void update(long fps) {
        switch (this.getFacing()) {
            case LEFT:
                x--;
                break;
            case RIGHT:
                x++;
                break;
            case UP:
                y--;
                break;
            case DOWN:
                y++;
                break;
        }
        super.move(fps);
    }

    public void setSpellType(String spellType, SpellObject spellObject) {
        switch (spellType) {
            case SpellNames.FIREBALL:
                SpellUtils.Companion.setFireball(spellObject);
                break;
        }
    }

    public void setWaitingTime(int waitingTime) {
        this.waitingTime = waitingTime;
    }

    public int getWaitingTime() {
        return waitingTime;
    }

    @Override
    public void setMoves(boolean moves) {
        super.setMoves(moves);
    }
}

File: app/src/main/java/com/iamagamedev/mybrandnewgame/gameObjects/spells/SpellObject.kt
package com.iamagamedev.mybrandnewgame.gameObjects.spells

import com.iamagamedev.mybrandnewgame.constants.SpellNames
import com.iamagamedev.mybrandnewgame.gameObjects.GameObject
import com.iamagamedev.mybrandnewgame.gameObjects.spells.SpellUtils.Companion.setFireball

abstract class SpellObject : GameObject() {
    var waitingTime = 3000
    private var x = 0f
    private var y = 0f

    override fun update(fps: Long) {
        when (facing) {
            LEFT -> x--
            RIGHT -> x++
            UP -> y--
            DOWN -> y++
        }
        super.move(fps)
    }

    open fun setSpellType(spellType: String?, spellObject: SpellObject?) {
        when (spellType) {
            SpellNames.FIREBALL -> setFireball(spellObject!!)
        }
    }
}